src/java/org/apache/cassandra/index/sai/plan/vector/VectorQueryController.java

```diff
@@ -0,0 +1,247 @@
+                                             .flatMap(expr -> expr.context.getMemtableIndexManager().iteratorsForSearch(queryContext, expr, mergeRange).stream())
+                                             .collect(Collectors.groupingBy(pair -> pair.left, Collectors.mapping(pair -> pair.right, Collectors.toList())));
+
+        VectorQueryViewBuilder.QueryView queryView = new VectorQueryViewBuilder(expressions, mergeRange).build();
+        // in case of ANN query in hybrid search, we have to reference ANN sstable indexes separately because queryView doesn't include ANN sstable indexes
+        VectorQueryViewBuilder.QueryView annQueryViewInHybridSearch = isAnnHybridSearch ? new VectorQueryViewBuilder(Collections.singleton(annExpressionInHybridSearch), mergeRange).build() : null;
+
+        try
+        {
+            List<KeyRangeIterator> sstableIntersections = new ArrayList<>(queryView.view.size());
+            for (Map.Entry<SSTableReader, List<VectorQueryViewBuilder.IndexExpression>> viewEntry : queryView.view.entrySet())
+            {
+                PostingList postings = createRowIdIterator(viewEntry.getValue(), isAnnHybridSearch);
+                if (isAnnHybridSearch)
+                    sstableIntersections.add(reorderAndLimitBySSTableRowIds(postings, viewEntry.getKey(), annQueryViewInHybridSearch));
+                else
+                {
+                    PrimaryKeyMap.Factory pkFactory = viewEntry.getValue().iterator().next().index.getSSTableContext().primaryKeyMapFactory;
+                    sstableIntersections.add(convertToPrimaryKeyIterator(pkFactory, postings));
+                }
+            }
+
+            List<KeyRangeIterator> memtableIntersections = iteratorsByMemtable.entrySet()
+                                                                              .stream()
+                                                                              .map(e -> {
+                                                                                  // we need to do all the intersections at the index level, or ordering won't work
+                                                                                  KeyRangeIterator it = buildIterator(e.getValue(), isAnnHybridSearch);
+                                                                                  if (isAnnHybridSearch)
+                                                                                      it = reorderAndLimitBy(it, e.getKey(), annExpressionInHybridSearch);
+                                                                                  return it;
+                                                                              })
+                                                                              .collect(Collectors.toList());
+
+            Iterable<KeyRangeIterator> allIntersections = Iterables.concat(sstableIntersections, memtableIntersections);
```

Comment:
I don't think this is correct.  Need to combine memtable results w/ sstable results at the expression level or you can miss matching rows that are split across multiple sources.

vsearch fix for this is 49a51690ca. it's a big diff though because you did a bunch of other refactoring here.

Reply:
This is the commit I'm intending bringing in. I believe this will address a number of issues.

Reply:
I didn't do an exhaustive comparison but it looks like we're still missing the tests that 49a51690 added in ComplexQueryTest, VectorTypeTest, and VectorUpdateDeleteTest.